### PR TITLE
Keep ZM coordinates when simplifying geometry

### LINKED ISSUE #50742: Area Simplify drops Z and M values
### What is the bug or the crash?

If you run the Simplify Algo with Area(Visvalingam) the Z and M values will become null if any vertexes are removed.
 

### Steps to reproduce the issue

1. Create a line layer with Z at 100 and M values set and enough vertices to be able to run simplify. 
2. Run the simplify algo and set Simplification method to Area(Visvalingam). 
3. Set the tolerance to a value that will remove at least one vertex. 
4. Check that the M and Z values are now null. 
Test data: 
[Area_Simplify_ZMdrop.zip](https://github.com/qgis/QGIS/files/9896735/Area_Simplify_ZMdrop.zip)

Starting layer with all Z set to 100 
image

Simplified with some vertex's removed all Z now null or 0 
image

Simplified with no vertex's removed all Z's still 100 (one vertex was set to 200 did the same thing without that) 
image
 

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.29.0-Master | QGIS code revision | ca130dd49c
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.9.5
GDAL/OGR version | 3.6.0dev-5df18be65c
PROJ version | 9.1.0
EPSG Registry database version | v10.074 (2022-08-01)
GEOS version | 3.10.3-CAPI-1.16.1
SQLite version | 3.39.4
PDAL version | 2.4.3
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.6
QScintilla2 version | 2.13.1
OS version | Windows 10 Version 2009
  |   |   |  
This copy of QGIS writes debugging output.
  |   |   |  
Active Python plugins
db_manager | 0.1.20
grassprovider | 2.12.99
MetaSearch | 0.3.6
otbprovider | 2.12.99
processing | 2.12.99
sagaprovider | 2.12.99

</body></html>QGIS version
3.29.0-Master
QGIS code revision
[ca130dd49c](https://github.com/qgis/QGIS/commit/ca130dd49c)
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.6.0dev-5df18be65c
PROJ version
9.1.0
EPSG Registry database version
v10.074 (2022-08-01)
GEOS version
3.10.3-CAPI-1.16.1
SQLite version
3.39.4
PDAL version
2.4.3
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009

This copy of QGIS writes debugging output.

Active Python plugins
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
otbprovider
2.12.99
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

The Distance (Douglas-Puecker) stetting does not appear to be affected but it also doesn't seem to take into account Z values in simplification.
I was excited by the prospect of being able to simplify draped lines and have them take into account the Z dimension which the Area setting does. 
I did not try this on polygons.

### PR DESCRIPTION
**Pull request by troopa81**
Fixes #50742

Fix Visvalingam and snaptogrid, not Douglas-Peucker distance method which remove M but not Z and rely on Geos.

### CODE REVIEW

File: src/core/qgsmaptopixelgeometrysimplifier.cpp [hunk 1]
```diff
@@ -202,11 +218,23 @@ std::unique_ptr< QgsAbstractGeometry > QgsMapToPixelSimplifier::simplifyGeometry
           {
             x = *xData++;
             y = *yData++;
+
+            if ( geometry.is3D() )
```
Comment (rouault) [inline src/core/qgsmaptopixelgeometrysimplifier.cpp:222]: should be a tiny bit faster:
```suggestion
            if ( zData )
```

File: src/core/qgsmaptopixelgeometrysimplifier.cpp [hunk 2]
```diff
@@ -202,11 +218,23 @@ std::unique_ptr< QgsAbstractGeometry > QgsMapToPixelSimplifier::simplifyGeometry
...
+              z = *zData++;
+
+            if ( geometry.isMeasure() )
```
Comment (rouault) [inline src/core/qgsmaptopixelgeometrysimplifier.cpp:225]:
```suggestion
            if ( mData )
```

**Review by rouault — APPROVED** [review]

File: src/core/qgsmaptopixelgeometrysimplifier.cpp (hunk 1 above)
Comment (troopa81) [reply src/core/qgsmaptopixelgeometrysimplifier.cpp]: Fixed with bool variable

**Comment by troopa81** [comment]
@rouault I treated your comments by using intermediary bool variables and fixed clazy  warning at the same time. Thanks for reviewing

**Pull request merged**